libraries/Microsoft.Bot.Builder.Azure/AzureTableStorage.cs

```diff
@@ -3,164 +3,369 @@
+
+                if (tableEntity.HttpStatusCode == (int)HttpStatusCode.OK)
                 {
-                    var value = ((StoreItemEntity)result.Result).AsObject();
-                    var valueStoreItem = value as IStoreItem;
-                    if (valueStoreItem != null)
-                    {
-                        valueStoreItem.eTag = result.Etag;
-                    }
-                    storeItems.Add(new KeyValuePair<string, object>(key, value));
+                    // re-create expected object
+                    StoreItemEntity storeItem = StoreItemEntity.AsStoreItemEntity(tableEntity);
+                    return new KeyValuePair<string, object>(key, storeItem.Entity);
                 }
+
+                return new KeyValuePair<string, object>();
+            });
+
+            return (await Task.WhenAll(readTasks).ConfigureAwait(false))
+                .Where(kv => kv.Key != null);
+        }
+
+        /// <summary>
+        /// Saves store items to storage.
+        /// </summary>
+        /// <param name="changes">Map of items to write to storage.</param>
+        /// <returns></returns>
+        public async Task Write(IEnumerable<KeyValuePair<string, object>> changes)
+        {
+            if (changes == null) throw new ArgumentNullException(nameof(changes));
+
+            var storeItems = changes.Select(kv => new StoreItemEntity(new EntityKey(kv.Key), kv.Value));
+            var bogusEtagKeys = storeItems.Where(item => item.ETag != null && item.ETag.Length == 0);
+            if (bogusEtagKeys.Any())
+            {
+                throw new ArgumentException("Bogus etag in items with key: " + string.Join(", ", bogusEtagKeys.Select(o => o.Key)));
             }
-            return storeItems;
+
+            await EnsureTableExists();
+
+            var writeTasks = changes.Select(kv =>
+            {
+                var storeEntity = new StoreItemEntity(new EntityKey(kv.Key), kv.Value);
+
+                // Re-create object as table entity
+                var tableEntity = storeEntity.AsTableEntity();
+
+                if (storeEntity.ETag == null || storeEntity.ETag == "*")
+                {
+                    // New item or etag=* then insert or replace unconditionaly
+                    return Table.ExecuteAsync(TableOperation.InsertOrReplace(tableEntity));
+                }
+
+
+                // Optimistic Update
+                return Table.ExecuteAsync(TableOperation.Replace(tableEntity));
+            });
+
+            await Task.WhenAll(writeTasks).ConfigureAwait(false);
         }
 
+        /// <summary>
+        /// Ensure table is created.
+        /// </summary>
+        /// <returns></returns>
+        private async Task EnsureTableExists()
+        {
+            if (!_tableCreated)
+            {
+                await Table.CreateIfNotExistsAsync();
+            }
+        }
```

Comment:
I would personally design this as `private ValueTask<CloudTable> GetTable()` and make it the factory for the table altogether. Axe the `Table` property (as mentioned above), axe the `_tableCreated` field, just check  `if(_table == null)` and create the `CloudTable` and check that it exists in there and then set the field. Subsequent calls just get the instance returned straight away.

Something like this (forgive any errors... just typing up in this comment):

```CSharp
private static async ValueTask<CloudTable> GetTable()
{
    if(_table != null)
    {
        return new ValueTask<CloudTable>(_table);
    }

    return new ValueTask<CloudTable>(EnsureTableExists());

    Task<CloudTable> EnsureTableExists()
    {
         var table = _storage.CloudTableClient().GetTableReference(_tableName);

         await table.CreateIfNotExistsAsync();

         _table = table;
    } 
}
```